`context.rewrite` broken in starlight 0.32.0

### What version of `starlight` are you using?

0.32.0

### What version of `astro` are you using?

5.3.0

### What package manager are you using?

npm

### What operating system are you using?

Linux (NixOS 24.11)

### What browser are you using?

Firefox 134.0.2

### Describe the Bug

Hi, me again 🌞 (look who's busy using starlight)

Using `context.rewrite` seems to have broken in the latest update from 0.31.1 to 0.32.0.
This example doesn't make much sense semantically but is enough to show the problem. Going off the starlight starter template (`npm create astro@latest -- --template starlight`) and adding the following in `src/middleware.ts`:

```ts
import { defineMiddleware } from "astro/middleware";

export const onRequest = defineMiddleware((context, next) => {
    if (context.params.slug && context.params.slug !== "/")
        return context.rewrite("/404");
    return next();
});
```

will behave just as you expect with astro 5.3.0 and starlight 0.31.1: The default splash page is shown on `/`, and on clicking the "Example Guide" button, the page shows the default 404 page with server logs something like this:

```
20:58:31 watching for file changes...
20:58:41 [200] / 123ms
20:58:41 [200] /_image 55ms
Entry docs → 404 was not found.
20:58:43 [404] (rewrite) /guides/example/ 20ms
```

when trying this with starlight 0.32.0, the landing page still works like before, but visiting a to-be-rewritten url results in these logs:

```
21:03:37 watching for file changes...
21:03:39 [200] / 122ms
21:03:40 [200] / 13ms
21:03:41 [ERROR] Cannot redefine property: starlightRoute
  Stack trace:
    at Function.defineProperty (<anonymous>)
    [...] See full stack trace in the browser, or rerun with --verbose.
21:03:41 [WARN] [router] A `getStaticPaths()` route pattern was matched, but no matching static path was found for requested path `/favicon.ico`.

Possible dynamic routes being matched: node_modules/@astrojs/starlight/routes/static/index.astro.
Entry docs → 404 was not found.
```

and a rather verbose error info in the browser that doesn't really help me much:

<details>
<summary>click to expand wall of text</summary>

```
Cannot redefine property: starlightRoute
```
`starlight/locals.ts:24:9`
```ts
import type { APIContext } from 'astro';
import { AstroError } from 'astro/errors';
import { defineMiddleware } from 'astro:middleware';
import type { StarlightRouteData } from './route-data';
import { useTranslations } from './utils/translations';

export const onRequest = defineMiddleware(async (context, next) => {
	context.locals.t = useTranslations(context.currentLocale);
	initializeStarlightRoute(context);
	return next();
});

/**
 * Sets up a `starlightRoute` property on locals. Initially, this will throw an error if accessed.
 * When rendering, Starlight’s routes set `starlightRoute` with the resolved route data object for
 * the current page.
 *
 * This ensures Starlight components can easily access `starlightRoute` without needing type guards,
 * we can throw a helpful message if `starlightRoute` is accessed on non-Starlight pages, and we
 * avoid generating route data in this middleware which also runs for non-Starlight route.
 */
export function initializeStarlightRoute(context: APIContext) {
	const state: { routeData: StarlightRouteData | undefined } = { routeData: undefined };
	Object.defineProperty(context.locals, 'starlightRoute', {
               ^
		get() {
			if (!state.routeData) {
				throw new AstroError(
					'`locals.starlightRoute` is not defined',
					'This usually means a component that accesses `locals.starlightRoute` is being rendered outside of a Starlight page, which is not supported.\n\n' +
						'If this is a component you authored, you can do one of the following:\n\n' +
						'1. Avoid using this component in non-Starlight pages.\n' +
						'2. Wrap the code that reads `locals.starlightRoute` in a  `try/catch` block and handle the cases where `starlightRoute` is not available.\n\n' +
						'If this is a Starlight built-in or third-party component, you may need to report a bug or avoid this use of the component.'
				);
			}
			return state.routeData;
		},
		set(routeData: StarlightRouteData) {
			state.routeData = routeData;
		},
	});
}

```
```
TypeError: Cannot redefine property: starlightRoute
    at Function.defineProperty (<anonymous>)
    at initializeStarlightRoute (/home/willy/repos/astro-starlight-middleware-rewrite-reproducer/node_modules/@astrojs/starlight/locals.ts:24:9)
    at /home/willy/repos/astro-starlight-middleware-rewrite-reproducer/node_modules/@astrojs/starlight/locals.ts:9:2
    at applyHandle (/home/willy/repos/astro-starlight-middleware-rewrite-reproducer/node_modules/astro/dist/core/middleware/sequence.js:20:22)
    at /home/willy/repos/astro-starlight-middleware-rewrite-reproducer/node_modules/astro/dist/core/middleware/sequence.js:17:12
    at applyHandle (file:///home/willy/repos/astro-starlight-middleware-rewrite-reproducer/node_modules/astro/dist/core/middleware/sequence.js:20:22)
    at file:///home/willy/repos/astro-starlight-middleware-rewrite-reproducer/node_modules/astro/dist/core/middleware/sequence.js:56:18
    at file:///home/willy/repos/astro-starlight-middleware-rewrite-reproducer/node_modules/astro/dist/i18n/middleware.js:49:28
    at applyHandle (file:///home/willy/repos/astro-starlight-middleware-rewrite-reproducer/node_modules/astro/dist/core/middleware/sequence.js:20:22)
    at file:///home/willy/repos/astro-starlight-middleware-rewrite-reproducer/node_modules/astro/dist/core/middleware/sequence.js:17:12
```
</details>

This was almost certainly broken by #2390, at least I recognize the name `starlightRoute` from there. As far as I can understand, the new "route middlewares" are _not_ what I want here since I'm rewriting on a "webserver level", not just changing page (route) data.

stackblitz sadly currently seems to be shitting itself, at least on my end, so all I have as a reproducer is the linked github repo. here are the steps to reproduce:

1. checkout
2. `npm ci`
3. `astro run dev`
4. click "Example Guide"
5. observe error

and additionally, to show that this worked before:

1. change starlight version from `0.32.0` to `0.31.1` in `package.json`
2. `npm i`
3. `npm run dev`
4. again, click "Example Guide"
5. see 404 page (hopefully)

Is this a bug, or am I just doing something wrong? (that just accidentally worked before) - I'd like to note that the log messages when rewriting to `/404` seem somewhat unexpected (404 not found while trying to show the 404 page? Although I might just be reading that wrong)

### Link to Minimal Reproducible Example

https://github.com/jemand771/astro-starlight-middleware-rewrite-reproducer

### Participation

- [ ] I am willing to submit a pull request for this issue.

--- a/.changeset/forty-ways-beam.md
+++ b/.changeset/forty-ways-beam.md
@@ -2,4 +2,4 @@
 '@astrojs/starlight': patch
 ---
 
-Fixes an issue preventing to use [rewrites](https://docs.astro.build/en/guides/routing/#rewrites).
+Fixes an issue preventing the use of [rewrites](https://docs.astro.build/en/guides/routing/#rewrites).

--- a/.changeset/forty-ways-beam.md
+++ b/.changeset/forty-ways-beam.md
@@ -1,0 +1,5 @@
+---
+'@astrojs/starlight': patch
+---
+
+Fixes an issue preventing to use [rewrites](https://docs.astro.build/en/guides/routing/#rewrites).

--- a/packages/starlight/__e2e__/fixtures/ssr/src/middleware.ts
+++ b/packages/starlight/__e2e__/fixtures/ssr/src/middleware.ts
@@ -1,0 +1,8 @@
+import { defineMiddleware } from 'astro:middleware';
+
+export const onRequest = defineMiddleware((context, next) => {
+	if (context.url.pathname === '/content') {
+		return context.rewrite('/demo');
+	}
+	return next();
+});

--- a/packages/starlight/__e2e__/ssr.test.ts
+++ b/packages/starlight/__e2e__/ssr.test.ts
@@ -54,6 +54,14 @@ test('SSR mode renders the same splash page as prerendering', async ({
 	expectEquivalentHTML(prerenderContent, ssrContent);
 });
 
+test('supports middleware rewriting', async ({ page, getProdServer }) => {
+	const starlight = await getProdServer();
+	const response = await starlight.goto('/content');
+
+	expect(response?.status()).toBe(200);
+	await expect(page.locator('#server-check')).toHaveText('On server');
+});
+
 function expectEquivalentHTML(a: string, b: string) {
 	expect(getNormalizedHTML(a)).toEqual(getNormalizedHTML(b));
 }

--- a/packages/starlight/locals.ts
+++ b/packages/starlight/locals.ts
@@ -20,8 +20,8 @@ export const onRequest = defineMiddleware(async (context, next) => {
  * avoid generating route data in this middleware which also runs for non-Starlight route.
  */
 export function initializeStarlightRoute(context: APIContext) {
-	const state: { routeData: StarlightRouteData | undefined } = { routeData: undefined };
 	if ('starlightRoute' in context.locals) return;
+	const state: { routeData: StarlightRouteData | undefined } = { routeData: undefined };
 	Object.defineProperty(context.locals, 'starlightRoute', {
 		get() {
 			if (!state.routeData) {

--- a/packages/starlight/locals.ts
+++ b/packages/starlight/locals.ts
@@ -21,6 +21,7 @@ export const onRequest = defineMiddleware(async (context, next) => {
  */
 export function initializeStarlightRoute(context: APIContext) {
 	const state: { routeData: StarlightRouteData | undefined } = { routeData: undefined };
+	if ('starlightRoute' in context.locals) return;
 	Object.defineProperty(context.locals, 'starlightRoute', {
 		get() {
 			if (!state.routeData) {

